File: apps/demo-fe-elixir-phoenix/lib/demo_fe_exph/api/attachments.ex
defmodule DemoFeExph.Api.Attachments do
  @moduledoc "Expense attachment API calls: list, upload, and delete."

  alias DemoFeExph.Api.Client

  @doc "Lists all attachments for a given expense."
  def list_attachments(token, expense_id) do
    Client.get("/api/v1/expenses/#{expense_id}/attachments", token)
  end

  @doc "Uploads a file attachment to the given expense."
  def upload_attachment(token, expense_id, file_path, content_type) do
    headers = Client.auth_headers(token)
    url = "#{Client.backend_url()}/api/v1/expenses/#{expense_id}/attachments"

    multipart =
      {:multipart,
       [
         {:file, file_path,
          {"form-data", [{"name", "file"}, {"filename", Path.basename(file_path)}]},
          [{"Content-Type", content_type}]}
       ]}

    case Req.post(url, body: multipart, headers: headers) do
      {:ok, %{status: status, body: body}} when status in 200..299 -> {:ok, body}
      {:ok, %{status: status, body: body}} -> {:error, {status, body}}
      {:error, reason} -> {:error, reason}
    end
  end

  @doc "Deletes an attachment from the given expense."
  def delete_attachment(token, expense_id, attachment_id) do
    Client.delete("/api/v1/expenses/#{expense_id}/attachments/#{attachment_id}", token)
  end
end

File: apps/demo-fe-ts-tanstackstart/app/lib/api/attachments.ts
import { apiFetch, getAccessToken } from "./client";
import type { Attachment } from "./types";

export function listAttachments(expenseId: string): Promise<Attachment[]> {
  return apiFetch<Attachment[]>(`/api/v1/expenses/${expenseId}/attachments`);
}

export async function uploadAttachment(expenseId: string, file: File): Promise<Attachment> {
  const formData = new FormData();
  formData.append("file", file);

  const token = getAccessToken();
  const headers: Record<string, string> = {};
  if (token) headers["Authorization"] = `Bearer ${token}`;

  const res = await fetch(`/api/v1/expenses/${expenseId}/attachments`, {
    method: "POST",
    headers,
    body: formData,
  });

  if (!res.ok) {
    const body = await res.json().catch(() => null);
    throw new Error(`Upload failed: ${res.status} ${JSON.stringify(body)}`);
  }

  return res.json() as Promise<Attachment>;
}

export function deleteAttachment(expenseId: string, attachmentId: string): Promise<void> {
  return apiFetch(`/api/v1/expenses/${expenseId}/attachments/${attachmentId}`, { method: "DELETE" });
}
